orttraining/orttraining/core/session/training_session.cc

```diff
@@ -1124,5 +1144,660 @@ std::unordered_set<std::string> TrainingSession::GetTrainableModelInitializers(
+              tensorboard_scalar_names.emplace_back(it->second);
+            }
+          };
+
+      add_opt_graph_output_by_key(OptimizerOutputKey::GradientAllIsFinite);
+      add_opt_graph_output_by_key(OptimizerOutputKey::GlobalGradientNorm);
+    }
+
+    ORT_RETURN_IF_ERROR(AddTensorboard(
+        tensorboard_config.summary_name, tensorboard_scalar_names,
+        tensorboard_config.histogram_node_names, tensorboard_config.norm_node_names,
+        tensorboard_config.dump_convergence_metrics));
+  }
+
+  // add GIST encoding
+  if (config.gist_config.has_value()) {
+    ORT_RETURN_IF_ERROR(AddGistEncoding());
+  }
+
+  // If the current node is in rank0 or if the current session is running pipeline (in which case different rank would
+  // store different model partition), and if model_with_training_graph_path is specified, save the model.
+  // Note: in the pipeline case, different ranks may resident in the same node. This could lead to a potential write
+  // conflict. It is user's responsibility to make sure different rank is passed in with different. Also, to avoid
+  // writing conflict, only the ranks in first pipeline group write the partition file out.
+  // model_with_training_graph_path value.
+  if ((IsRootNode(config) || (config.pipeline_config.has_value() &&
+                              DistributedRunContext::GroupId(WorkerGroupType::PipelineParallel) == 0)) &&
+      config.model_with_training_graph_path.has_value()) {
+    ORT_IGNORE_RETURN_VALUE(Save(
+        config.model_with_training_graph_path.value(), SaveOption::NO_RELOAD));
+  }
+
+  // After pipeline partition, we need to return the inputs allowed in this partition.
+  if (config.pipeline_config.has_value()) {
+    const auto& allowed_inputs = model_->MainGraph().GetInputsIncludingInitializers();
+    const auto& allowed_outputs = model_->MainGraph().GetInputsIncludingInitializers();
+    for (size_t i = 0; i < allowed_inputs.size(); ++i) {
+      const auto name = allowed_inputs[i]->Name();
+      config_result.pipeline_config_result.value().feed_names.push_back(name);
+    }
+    for (size_t i = 0; i < allowed_outputs.size(); ++i) {
+      const auto name = allowed_outputs[i]->Name();
+      config_result.pipeline_config_result.value().fetch_names.push_back(name);
+    }
+
+#if defined(USE_NCCL) && defined(USE_NCCL_P2P)
+    LaunchNcclService(config_result.pipeline_config_result.value().pipeline_stage_id);
+#endif
+  } else {
+#if defined(USE_NCCL) && defined(USE_NCCL_P2P)
+    LaunchNcclService(pipeline_context_.pipeline_stage_id);
+#endif
+  }
+
+  config_result_out = std::move(config_result);
+  is_configured_ = true;
+
+  return Status::OK();
+}
+
+common::Status PipelineTrainingSession::Run(const RunOptions& run_options, IOBinding& io_binding) {
+  if (pipeline_context_.num_pipeline_stages > 1) {
+    return RunWithPipeline(run_options, io_binding);
+  } else {
+    return RunWithoutPipeline(run_options, io_binding);
```

Comment:
>return RunWithoutPipeline(run_options, io_binding); [](start = 4, length = 51)

If no pipeline is involved, maybe it is a mistake to trigger `PipelineTrainingSession`. Does it make more sense to error out here?

Reply:
Humm, probably no. Python API is typed to PipelineTrainingSession now. Even if `pipeline_parallel_size=1`, the `session_` still need a valid `Run(...)`.

---
In reply to: [538816596](https://github.com/microsoft/onnxruntime/pull/5815#discussion_r538816596) [](ancestors = 538816596)